### PR TITLE
Initial multi-column pooler

### PR DESCRIPTION
**Pull request by subutai**
with basic unit tests

### CODE REVIEW

**Comment by subutai** [comment]
@natoromano If you can do a quick scan before I merge that would be great. Thanks!

File: htmresearch/algorithms/column_pooler.py [hunk 1]
```diff
@@ -110,19 +115,17 @@ def compute(self,
...
+    @param learn                    (bool)
+            If True, we are learning a new object
 
     """
     if activeExternalCells is None:
-      activeExternalCells = set()
-
-    if activeApicalCells is None:
-      activeApicalCells = set()
+      activeExternalCells = []
```
Comment (natoromano) [inline htmresearch/algorithms/column_pooler.py]: This should clearly documented somewhere in NuPIC. It caused slight problems in my experiments. Some regions/algorithms are using `set`s, whereas some other use `list`s.

File: htmresearch/algorithms/column_pooler.py [hunk 2]
```diff
@@ -181,15 +180,16 @@ def _computeLearningMode(self, feedforwardInput, lateralInput):
...
+                      activeExternalCells=lateralInput,
+                      formInternalConnections=False,
+                      learn=True)
 
 
   def _computeInferenceMode(self, feedforwardInput, lateralInput):
     """
     Inference mode: if there is some feedforward activity, perform
     spatial pooling on it to recognize previously known objects. If there
-    is no feedforward activity, maintain previous activity.
```
Comment (natoromano) [inline htmresearch/algorithms/column_pooler.py]: You may have deleted a line by inadvertence.

File: htmresearch/algorithms/column_pooler.py (hunk 1 above)
Comment (mrcslws) [inline htmresearch/algorithms/column_pooler.py]: FWIW the new TM/ETM doesn't use sets, so this awkwardness should go away after we finish updating the Python ETM.

File: htmresearch/algorithms/column_pooler.py (hunk 2 above)
Comment (subutai) [inline htmresearch/algorithms/column_pooler.py]: Thanks - good catch

File: htmresearch/algorithms/column_pooler.py (hunk 1 above)
Comment (subutai) [inline htmresearch/algorithms/column_pooler.py]: OK. In this implementation this line should really assign set(). I'll fix it.

File: htmresearch/algorithms/temporal_memory_phases.py [hunk 3]
```diff
@@ -350,7 +350,9 @@ def learnOnSegments(self,
                           self.permanenceDecrement)
 
       if isLearningSegment:
-        n = self.maxNewSynapseCount - len(activeSynapses)
+        n = min(self.maxNewSynapseCount,
```
Comment (subutai) [inline htmresearch/algorithms/temporal_memory_phases.py]: @mrcslws Note that I did make the adding synapses bug fix in this implementation, as it is needed to form unions.

**Comment by mrcslws** [comment]
> @mrcslws Note that I did make the adding synapses bug fix in this implementation, as it is needed to form unions.

Ok. This might change the output of the `faulty_temporal_memory`. I don't know if this approach has been tested before, considering the TP had the synapses bug. But it seems right.

**Pull request merged**